### PR TITLE
RFC : CDEvents Webhook Adapter design review

### PR DESCRIPTION
**Pull request by rjalander**
Creating this PR, to review the current design approach and propose any changes that might help in creating a common cdevents-translator library.

Adding a page for Gerrit-CDEvents integration and mapping for various Gerrit events to CDEvents.

### CODE REVIEW

File: rfc/cdevents-translator.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
+## CDEvents Translator
+
+### Overview:
+This proposal describes translating events from various source control systems into CDEvents and sending them to configured message-broker, by using a common cdevents-translator 
+
+A library cdevents-translator will be developed to receive various event types and convert them into CDEvents.
```
Comment (afrittoli) [inline rfc/cdevents-translator.md:6]: How this will work in terms of supporting multiple types?
Will the library provide an interface that various implementors can implement independently?

File: rfc/cdevents-translator.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+Various types of source code version control systems like GitHub, Gerrit and Gitlab endpoints/webhooks can be configured and implemented with in the cdevents-translator.
+
+Instructions for configuring Gerrit webhooks and implementation details can be found in [gerrit-cdevent](gerrit-cdevents.md) 
+
+### Design approach
+Create a Golang application with an API server configured to initialize SCM systems webhooks/endpoints to receive events and translate them into CDEvents.
```
Comment (afrittoli) [inline rfc/cdevents-translator.md:13]: The following just for info.

Tekton Triggers provides and API server that can reply to webhooks. It supports incoming `CloudEvents` as well as generic webhooks and it provides a powerful templating mechanism to extract information from incoming payloads and produce a YAML out of them. I used it in this POC https://github.com/afrittoli/cdeventer.

The approach of CDEventer is a bit more complex, but it has the advantage of decoupling the processing of incoming messages, which is then implemented as a set of input specific `TriggerBindings`, from the actual sending of the target CDEvents, through the use of an intermediate YAML.

The downside of CDEventer is the extra dependency to Tekton (and thus K8s).

File: rfc/cdevents-translator.md (hunk 1 above)
Comment (rjalander) [reply rfc/cdevents-translator.md:6]: yes an interface `EventTranslator` will be created with the implementation needed for `TranslateEvent()` for different source of events(GitHub/Gerrit`/...etc)

````go
type EventTranslator interface {
	TranslateEvent()
}
````
And will have a common methods to create and send CDEvents, will update the design as per this.

**Review by xibz — CHANGES_REQUESTED** [review]

File: rfc/cdevents-translator.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
+## CDEvents Translator
+
+### Overview:
```
[the remark this answers is not in the document]
Comment (rjalander) [reply rfc/cdevents-translator.md:3]: Thank you @xibz for reviewing this and sharing your approaches.
As discussed over call, looking into these proposals and trying to explore a bit to update the design doc with more details on these approaches.

[the remark this answers is not in the document]
Comment (rjalander) [reply rfc/cdevents-translator.md:3]: Updated the design as per the discussion with different approaches. Please review and share your comments. 
we can meet and discuss further.

**Comment by rjalander** [comment]
Thank you @xibz for directing various approaches for the translator.

@afrittoli @xibz please let us know your views on these design proposals and review.

File: rfc/cdevents-translator.md [hunk 4]
```diff
@@ -1,65 +1,225 @@
...
+}
 ````
-An interface `EventTranslator` will be created to handle `TranslateEvent`. 
-Different source of systems need to implement `TranslateEvent()` to handle translation from source events to CDEvents.
 
+#### Pros and Cons
+This structure can be extended by adding more translator libraries in the future without modifying the main application code.</br>
+The problem with this approach anyone who is wanting some custom set of translator would need to write their own main class.
+
+### Approach 2 : Creating a Go's plugin system to implement different Translators
```
Comment (xibz) [inline rfc/cdevents-translator.md:62]: I'd probably rename this to "Using Go plugins to implement different Translators"

File: rfc/cdevents-translator.md [hunk 5]
```diff
@@ -1,65 +1,225 @@
...
+
+#### Pros and Cons
+
+This structure can be extended by adding more translator plugins in the future without having to recompile the entire application.</br>
+But the Go plugin system is currently only supported on Unix-like systems and has some limitations, such as the need to compile everything with the same Go version.</br>
+Expose the functionality of Go translator plugins as an HTTP/REST API, if this needs be used from other languages.
 
-For example the event structure is different for various events produced by Gerrit and an example struct of Gerrit project created event as below,
+
+### Approach 3 : Creating Go Plugin System over RPC(gRPC) using HashiCorp's go-plugin
```
Comment (xibz) [inline rfc/cdevents-translator.md:110]: "Using RPC to add new Translators"

File: rfc/cdevents-translator.md [hunk 6]
```diff
@@ -1,65 +1,225 @@
...
-	if err != nil {
-		fmt.Println("Error occurred while Unmarshal gerritEvent data into ProjectCreated struct", err)
-		return nil, err
-	}
-	return cdevent, nil
+package main
+
+
+type GerritTranslator struct {
+   //Fields for Gerrit event Translator
```
Comment (xibz) [inline rfc/cdevents-translator.md:145]: nit: `// [.]+`

File: rfc/cdevents-translator.md [hunk 7]
```diff
@@ -1,65 +1,225 @@
...
+
+This structure can be extended by adding more translator plugins in the future without having to recompile the entire application.</br>
+But the Go plugin system is currently only supported on Unix-like systems and has some limitations, such as the need to compile everything with the same Go version.</br>
+Expose the functionality of Go translator plugins as an HTTP/REST API, if this needs be used from other languages.
 
-For example the event structure is different for various events produced by Gerrit and an example struct of Gerrit project created event as below,
+
+### Approach 3 : Creating Go Plugin System over RPC(gRPC) using HashiCorp's go-plugin
+ 
+HashiCorp's [`go-plugin`](https://github.com/hashicorp/go-plugin) library simplifies the implementation of a plugin system in Go, it is a Go (golang) plugin system over RPC created by HashiCorp.</br>
```
Comment (xibz) [inline rfc/cdevents-translator.md:112]: `plugin system in Go. It is a`

File: rfc/cdevents-translator.md [hunk 8]
```diff
@@ -109,9 +109,10 @@ func main() {
 
 #### Pros and Cons
 
-This structure can be extended by adding more translator plugins in the future without having to recompile the entire application.</br>
-But the Go plugin system is currently only supported on Unix-like systems and has some limitations, such as the need to compile everything with the same Go version.</br>
-Expose the functionality of Go translator plugins as an HTTP/REST API, if this needs be used from other languages.
+- This structure can be extended by adding more translator plugins in the future without having to recompile the entire application.
+- The plugins can be developed independently and reside in separate repositories.
+- But the Go plugin system is currently only supported on Unix-like systems and has some limitations, such as the need to compile everything with the same Go version.
```
Comment (xibz) [inline rfc/cdevents-translator.md:106]: I would remove the initial "But"

**Review by xibz — CHANGES_REQUESTED** [review]
This looks much better @rjalander! One thing I noticed missing was a final "Conclusions" to summarize which was preferred by you, etc. Or did you want to leave that up for discussion?

File: rfc/cdevents-translator.md (hunk 5 above)
Comment (rjalander) [reply rfc/cdevents-translator.md:110]: updated

**Comment by rjalander** [comment]
> Conclusions

Thank you @xibz for reviewing this PR,

Yes I think we can discuss in CDEvents Working Group meeting and reach a conclusion to select the most suitable one

File: rfc/cdevents-translator.md [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+````
+
+The GRPC client/server implementations can be [generated in other languages](https://protobuf.dev/getting-started/), to enable Cross-language support for Plugins implemented.</br>
+Plugin users use `plugin.Client` to launch a subprocess and request an interface implementation using hashicorp's go-plugin over RPC.</br>
+More detailed implementation can be referred from HashiCorp's go-plugin [gRPC examples](https://github.com/hashicorp/go-plugin/tree/main/examples/grpc) 
+
+#### Pros and Cons
+- This structure allows you to have a main application that dynamically loads and communicates with different translator plugins over gRPC.
+- The plugins can be developed independently and reside in separate repositories.</br>
+- HashiCorp's `go-plugin` library developed under Mozilla Public License.
```
Comment (afrittoli) [inline rfc/cdevents-translator.md:221]: Is this a pro or a cons? The Mozilla license is recognised by OSI https://opensource.org/license/mpl-2-0/ - do you see an issue with it?

File: rfc/cdevents-translator.md [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+
+````
+
+The GRPC client/server implementations can be [generated in other languages](https://protobuf.dev/getting-started/), to enable Cross-language support for Plugins implemented.</br>
+Plugin users use `plugin.Client` to launch a subprocess and request an interface implementation using hashicorp's go-plugin over RPC.</br>
+More detailed implementation can be referred from HashiCorp's go-plugin [gRPC examples](https://github.com/hashicorp/go-plugin/tree/main/examples/grpc) 
+
+#### Pros and Cons
+- This structure allows you to have a main application that dynamically loads and communicates with different translator plugins over gRPC.
+- The plugins can be developed independently and reside in separate repositories.</br>
```
Comment (afrittoli) [inline rfc/cdevents-translator.md:220]: If I'm not mistaken, this is the same for all options, so while it's a pro, it's not a differentiator...

NIT: </br> at the end?

**Comment by e-backmark-ericsson** [comment]
Some general comments on this RFC:

- A basic principle in CDEvents should be that the events are sent by their source itself, or as close to that source as possible. I understand that it will be challenging to get all tools (SCM tools in this case mostly) to adapt sending CDEvents, so we might need something to fill that gap. At least shor/mid-term. It would be good to see some statement about that in the RFC.
- The RFC seems to propose a one-directional translation only, from XYZ to CDEvents. Are there ideas on translating events the other way around as well? I believe it would be a valid limitation to only support one-way translation, and thus that can be clearly stated in the introduction to the RFC. Translating from CDEvents to XYZ could require a much more complex set up, depending on what downstream systems/tools to support.
- Could/should we limit what interface protocols this translator would support towards the SCM (or other) systems? Most (all?) systems support webhooks, so maybe we should limit to that? Gerrit also supports their own 'stream-events' over ssh (and also Kafka through a plugin), but I don't think we need to support that as the webhooks support is there..
- I'd like to see a clearer scope of the intended solution. 99% of it talks about SCM related events. Could/should we limit this service to only manage SCM events, or should it be kept open for other types of CDEvents as well?
- Based on my reflections above, I'm not sure about the term 'cdevents-translator' and 'translations' here. I believe the service should have a name that reflect better the intended scope of it. Some suggestions: 'scm_event_producer', 'scm_adapter', 'any_to_cdevents_adapter', 'webhook_cdevents_producer'
- It will be easier to grasp the proposal if a simple picture was included, showing where services are deployed and their interfaces to the SCM system(s) and event brokers etc.
- A consumer of these events will eventually want to provide feedback on the source change in the SCM system, so it would be great to see that included in use cases and in a schematic picture of this proposal. I don't expect this service itself to manage that feedback, or is that considered as an option as well?

Bottom line: I propose that this initiative is limited to manage SCM related events only, and that it is one-directional. The name should preferably in some way reflect that.

File: rfc/cdevents-translator.md [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+````
+
+The GRPC client/server implementations can be [generated in other languages](https://protobuf.dev/getting-started/), to enable Cross-language support for Plugins implemented.</br>
+Plugin users use `plugin.Client` to launch a subprocess and request an interface implementation using hashicorp's go-plugin over RPC.</br>
+More detailed implementation can be referred from HashiCorp's go-plugin [gRPC examples](https://github.com/hashicorp/go-plugin/tree/main/examples/grpc) 
+
+#### Pros and Cons
+- This structure allows you to have a main application that dynamically loads and communicates with different translator plugins over gRPC.
```
[the remark this answers is not in the document]
Comment (xibz) [reply rfc/cdevents-translator.md:219]: Kind of. It really depends on the design. For instance, how hashicorp does this is they have the main translator start all the grpc, e.g. forking processes.

File: rfc/cdevents-translator.md (hunk 6 above)
Comment (xibz) [reply rfc/cdevents-translator.md:145]:
```suggestion
   // Fields for Gerrit event Translator
```

File: rfc/cdevents-translator.md [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+````go
+package main
+
+
+type GerritTranslator struct {
+   //Fields for Gerrit event Translator
+}
+
+func (translator *GerritTranslator) TranslateEvent(event string) {
+	//handle translate Gerrit event to CDEvent
```
Comment (xibz) [inline rfc/cdevents-translator.md:149]:
```suggestion
	// handle translate Gerrit event to CDEvent
```

File: rfc/cdevents-translator.md (hunk 10 above)
Comment (rjalander) [reply rfc/cdevents-translator.md:220]: Agree, removing it from here and also it is mentioned in the each approach description.

**Comment by rjalander** [comment]
> Some general comments on this RFC:
> 
> * A basic principle in CDEvents should be that the events are sent by their source itself, or as close to that source as possible. I understand that it will be challenging to get all tools (SCM tools in this case mostly) to adapt sending CDEvents, so we might need something to fill that gap. At least shor/mid-term. It would be good to see some statement about that in the RFC.
> * The RFC seems to propose a one-directional translation only, from XYZ to CDEvents. Are there ideas on translating events the other way around as well? I believe it would be a valid limitation to only support one-way translation, and thus that can be clearly stated in the introduction to the RFC. Translating from CDEvents to XYZ could require a much more complex set up, depending on what downstream systems/tools to support.
> * Could/should we limit what interface protocols this translator would support towards the SCM (or other) systems? Most (all?) systems support webhooks, so maybe we should limit to that? Gerrit also supports their own 'stream-events' over ssh (and also Kafka through a plugin), but I don't think we need to support that as the webhooks support is there..
> * I'd like to see a clearer scope of the intended solution. 99% of it talks about SCM related events. Could/should we limit this service to only manage SCM events, or should it be kept open for other types of CDEvents as well?
> * Based on my reflections above, I'm not sure about the term 'cdevents-translator' and 'translations' here. I believe the service should have a name that reflect better the intended scope of it. Some suggestions: 'scm_event_producer', 'scm_adapter', 'any_to_cdevents_adapter', 'webhook_cdevents_producer'
> * It will be easier to grasp the proposal if a simple picture was included, showing where services are deployed and their interfaces to the SCM system(s) and event brokers etc.
> * A consumer of these events will eventually want to provide feedback on the source change in the SCM system, so it would be great to see that included in use cases and in a schematic picture of this proposal. I don't expect this service itself to manage that feedback, or is that considered as an option as well?
> 
> Bottom line: I propose that this initiative is limited to manage SCM related events only, and that it is one-directional. The name should preferably in some way reflect that.

@e-backmark-ericsson Thank you very much for your comments,
- Updated the design including the first 4 comments as part of Goals, Non-Goals and Motivation and Rationale
- And about the name change we discussed in the CDEvents Workgroup meeting yesterday, is that fine to change as `webhook_cdevents_translator` or `webhook_cdevents_adapter`, please comment.
- Will try to add a picture for this proposal as suggested
- I think a feedback from the consumer will be provided back to the message-broker with an another CDEvent, but the event will not be translated/send to SCM system, as stated above this translator supports one-way translation.

File: rfc/cdevents-translator.md [hunk 13]
```diff
@@ -1,21 +1,38 @@
...
+### Goals
+
+- The translators will be implemented for tools that support sending events over the HTTP protocol with webhook capabilities
+- Implement support for one-way translation only; that is, translating events from different SCM systems to CDEvents
+- The translators will be implemented to translate only [Source Code Control](https://cdevents.dev/docs/source-code-version-control/) type CDEvents
+- Create shared packages/interfaces that will be commonly used for translating and sending CDEvents. 
+- Finalize one of the three approaches and utilize the `cdevents-translator` library to implement the Gerrit translator initially.
+
+### Non-Goals
+- Implement translators for other SCM systems or any applications requiring translation to CDEvents
```
Comment (e-backmark-ericsson) [inline rfc/cdevents-translator.md:26]: I thought this proposal intends to potentially provide support for any SCM system, but no other system
```suggestion
- Implement translators for other applications (non-SCM systems) requiring translation to CDEvents
```

**Review by e-backmark-ericsson — COMMENTED** [review]
To me it makes sense to rename this proposed implementation to `webhook_cdevents_adapter`. A `translator` to me implies that there would be a one-to-one translation, but that might not always be the case I think.

**Comment by rjalander** [comment]
@e-backmark-ericsson @afrittoli @xibz Thank you all for reviewing and providing your comments.
Please approve this RFC If no other comments.

**Comment by rjalander** [comment]
> Yes I requested @afrittoli to create project under cdevents org with the name "webhook-cdevents-adapter" (replaced _ with -), just to follow other projects naming conventions under cdevents org
> 
> And about using `translator`, referring plugin implementation projects for various SCM tools as `<scm_tool_name>-translator-cdevents`(can be a separate project under different org.)
> 
> Ex: `gerrit-translator-cdevents` and the interface name `EventTranslator.TranslateEvent()` to be implemented.
> 
> Please suggest, If I can rename this too with `gerrit-adaptor-cdevents` and the interface name `EventAdaptor.TranslateEvent()` to implement ?

As discussed in one of the SIG event meets, keeping the plugin naming with `translator` word and using adapter for CDEvents webhook.

File: rfc/cdevents-translator.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (afrittoli) [reply rfc/cdevents-translator.md:3]: ok to resolve this one now?

**Comment by rjalander** [comment]
@afrittoli @xibz can you please re-review and approve If no other comments with this RFC.

**Review by afrittoli — APPROVED** [review]
Thanks @rjalander - lgtm!
Commits need to squashed - we can do a squash and merge too.
